pkg/policy/api.go

```diff
@@ -330,21 +330,15 @@ type RelationsCache struct {
 
 func NewBuiltins(
 	input *models.State,
-	resourcesResolver ResourcesResolver,
+	resourcesQuery *ResourcesQueryCache,
 	relations *RelationsCache,
 ) *Builtins {
-	// Share the same calledWith map across resource-querying builtins, so that
-	// all queried resources are returned by inputResourceTypes
-	inputResolver := newInputResolver(input)
-	resolver := ResourcesResolver(inputResolver.resolve)
-	if resourcesResolver != nil {
-		resolver = resolver.Or(resourcesResolver)
-	}
+	resourceTypesQueried := map[string]struct{}{}
 
 	return &Builtins{
-		resourcesQueried: inputResolver.calledWith,
+		resourceTypesQueried: resourceTypesQueried,
 		funcs: []builtin{
-			&Query{ResourcesResolver: resolver},
+			resourcesQuery.trackResourceTypes(resourceTypesQueried),
```

Comment:
Replying to the other comment: we pass in this `resourceTypesQueried` map here, but it's also put in the `resourceTypesQueried` field above -- that's where it is read out from (since maps are always passed by reference).

Reply:
Oh okay, now I see. The `Builtins` type is actually responsible for reading out these cached results!

Reply:
Yeah, and that’s maybe not the best place to do it… but that’s a cleanup for another day 🙂